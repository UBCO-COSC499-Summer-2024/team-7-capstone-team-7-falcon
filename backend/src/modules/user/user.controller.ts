import { Controller, HttpStatus, ParseIntPipe } from '@nestjs/common';
import { Get, Param, Res } from '@nestjs/common/decorators';
import { Response } from 'express';
import { UserService } from './user.service';
import { ERROR_MESSAGES } from '../../common';

@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}

  @Get('/')
  get(@Res() res: Response) {
    return res.status(HttpStatus.OK).send({
      message: 'ok',
    });
  }

  /**
   * Get user by id
   * @param res {Response} - Response object
   * @param uid {number} - User id
   * @returns {Promise<Response>} - User object
   */
  @Get('/:uid')
  async getById(
    @Res() res: Response,
    @Param('uid', ParseIntPipe) uid: number,
  ): Promise<Response> {
    const user = await this.userService.getUserById(uid);

    if (!user) {
      return res.status(HttpStatus.NOT_FOUND).send({
        message: ERROR_MESSAGES.userController.userNotFound,
      });
    } else {
      return res.status(HttpStatus.OK).send(user);
    }
  }
}
