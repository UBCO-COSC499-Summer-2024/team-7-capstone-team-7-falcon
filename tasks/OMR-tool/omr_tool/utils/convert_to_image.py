from pdf2image import convert_from_path
import logging
logger = logging.getLogger(__name__)


def convert_to_image(pdf_path, output_path, num_pages_per_sheet):
    try:
        images = convert_from_path(pdf_path)
        try:
            i = 1
            submission_num = 1
            for image in images:
                if i > num_pages_per_sheet:
                    i = 1
                    submission_num += 1
                image.save(f"{output_path}/submission_{submission_num}-page_{i}.jpg", "JPEG")
                i += 1
            
        except (FileNotFoundError, PermissionError, OSError) as e:
            logging.exception(f'Error saving images:\n {str(e)}')
            return 1
    except Exception as e:
        logging.exception(f'Error converting PDF to image:\n {str(e)}')
        return 2
    return 0

if __name__ == "__main__":
    pdf_path = "sample_sheets/20240619103158362.pdf"
    output_path = "sample_sheets/image_output"
    if convert_to_image(pdf_path, output_path, 2) == 0:
        print("PDF converted to image successfully")
    else:
        print("PDF to image conversion failed")