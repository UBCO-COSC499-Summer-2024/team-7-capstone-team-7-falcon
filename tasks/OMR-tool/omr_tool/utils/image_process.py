import cv2
import numpy as np
import sys
from pathlib import Path
from PIL import Image

# ALIGNMENT_TEMPLATE = cv2.imread(
#     Path(__file__).resolve().parents[2] / "fixtures" / "template" / "alignment.png",
#     cv2.COLOR_BGR2GRAY,
# )


def prepare_img(image: Image) -> Image:
    """
    Function to preprocess an image for processing.

    Args:
        image (PIL.Image): The image to be prepared.

    Returns:
        PIL.Image: The prepared image.

    """
    grayscale_img = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    portrait_img = align_img(grayscale_img)
    return portrait_img


def align_img(image: Image) -> Image:
    """
    Function to align an image.

    TODO: Implement a more robust alignment method.

    Args:
        image (PIL.Image): The image to be aligned.

    Returns:
        PIL.Image: The aligned image.

    """
    ogH, ogW = image.shape[:2]

    if ogW > ogH:
        image = cv2.rotate(image, cv2.ROTATE_90_COUNTERCLOCKWISE)
    return image


def threshold_img(image):
    """
    Function to threshold an image.

    Args:
        image (PIL.Image): The image to be thresholded.

    Returns:
        PIL.Image: The thresholded image.

    """
    initial_thresh = cv2.threshold(image, 225, 255, cv2.THRESH_BINARY)[1]
    blur_image = cv2.GaussianBlur(initial_thresh, (3, 3), 0)
    thresh = cv2.threshold(blur_image, 0, 255, cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)[
        1
    ]
    return thresh


def sort_contours(cnts):
    """
    Function to sort contours from top to bottom and left to right.

    Args:
        cnts (list): A list of contours to sort.

    Returns:
        list
    """

    # Sort the contours by y-value
    sorted_contours = sorted(cnts, key=lambda cnt: cv2.boundingRect(cnt)[1])  # y values

    return sorted_contours


def identify_object_contours(generated_contours):
    """
    CURRENTLY UNUSED
    Function to identify objects in an image without using machine learning.

    Args:
        contours (list): A list of contours to identify objects in.

    Returns:
        list: A list of objects identified in the image.
    """
    objects = []
    questionCnts = sort_contours(generated_contours)
    row_contours = []
    last_cnt_y = None
    for cnt in questionCnts:
        (x, y, w, h) = cv2.boundingRect(cnt)

        if last_cnt_y is None:
            row_contours.append(cnt)
            last_cnt_y = y

            continue

        if abs(y - last_cnt_y) < h:
            row_contours.append(cnt)
            last_cnt_y = y
            continue

        else:
            sorted_row = sorted(
                row_contours, key=lambda cnt: cv2.boundingRect(cnt)[0]
            )  # x values
            separated_row = []
            last_cnt_x = None
            for row_cnt in sorted_row:
                x, y, w, h = cv2.boundingRect(row_cnt)
                if last_cnt_x is None:
                    separated_row = [row_cnt]
                    last_cnt_x = x
                    continue
                if abs(x - last_cnt_x) < w:
                    separated_row.append(row_cnt)
                    last_cnt_x = x
                    continue
                else:
                    separated_row.append(row_cnt)
                    row_length = len(separated_row)
                    if row_length <= 5:
                        objects.append({"type": "question", "contours": separated_row})
                    elif row_length == 10:
                        objects.append({"type": "SN_digit", "contours": separated_row})
                    elif row_length > 20:
                        objects.append({"type": "name_char", "contours": separated_row})
                    else:
                        objects.append({"type": "name_char", "contours": separated_row})
                    separated_row = [row_cnt]
                last_cnt_x = x
            row_contours = [cnt]
            last_cnt_y = y

    # Add the last set of row_contours if needed
    if row_contours:
        row_length = len(row_contours)
        if row_length <= 5:
            objects.append({"type": "question", "contours": row_contours})
        elif row_length == 10:
            objects.append({"type": "SN_digit", "contours": row_contours})
        elif row_length > 20:
            objects.append({"type": "name_char", "contours": row_contours})
        else:
            objects.append({"type": "name_char", "contours": row_contours})
    return objects


def identify_bubbled(img, cnts):
    """
    Identifies the filled-in bubbles in an image.

    Args:
        img (numpy.ndarray): The input image.
        cnts (list): List of contours representing the bubbles.

    Returns:
        list: List of contours representing the filled-in bubbles.
    """
    bubbled = None
    filled_in = []

    for cnt, i in enumerate(cnts):
        mask = np.zeros(img.shape, dtype="uint8")
        cv2.drawContours(mask, [i], -1, 255, -1)

        mask = cv2.bitwise_and(img, img, mask=mask)
        total = cv2.countNonZero(mask)

        if bubbled is None or total > 400:
            bubbled = (total, cnt)
            filled_in.append(cnts[bubbled[1]])
    return filled_in


def draw_bubble_contours(image, bubble_contour, question_bounds, color, offset=(0, 0)):
    """
    Draw the bubble contours on the image.

    Args:
        image (numpy.ndarray): The input image.
        bubble_contours (list): The list of bubble contours.
        question_bounds (tuple): The bounding box coordinates (x1, y1, x2, y2) of the question region.

    Returns:
        numpy.ndarray: The image with the bubble contours drawn.

    """
    output_image = image.copy()
    x1, y1, x2, y2 = question_bounds
    repositioned_cnt = bubble_contour + [x1, y1]
    cv2.drawContours(output_image, [repositioned_cnt], -1, color, 2, offset=offset)
    return output_image


def highlight_error_region(image, question_bounds, message=""):
    output_image = image.copy()
    x1, y1, x2, y2 = question_bounds
    cv2.rectangle(output_image, (x1, y1), (x2, y2), (0, 0, 255), 2)
    if message:
        cv2.putText(
            output_image,
            message,
            (x1, y2 + 20),
            cv2.FONT_HERSHEY_SIMPLEX,
            1,
            (0, 0, 255),
            2,
        )
    return output_image


def extract_roi(image, question_bounds):
    """
    Extracts the region of interest (ROI) from the given image based on the provided question bounds.

    Parameters:
        image (numpy.ndarray): The input image from which the ROI needs to be extracted.
        question_bounds (tuple): The bounding box coordinates (x1, y1, x2, y2) of the question region.

    Returns:
        numpy.ndarray: The cropped ROI image.

    """
    mask = np.zeros(image.shape[:2], dtype="uint8")
    x1, y1, x2, y2 = question_bounds
    cv2.fillPoly(mask, [np.array([(x1, y1), (x2, y1), (x2, y2), (x1, y2)])], 255)
    roi = cv2.bitwise_and(image, image, mask=mask)
    roi_cropped = roi[y1:y2, x1:x2]
    return roi_cropped


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(
            "Please provide an image path as an argument, falling back to placeholder."
        )
        image_path = (
            Path(__file__).resolve().parents[2] / "fixtures" / "submission_2-page_1.jpg"
        )
    else:
        image_path = sys.argv[1]
    image = cv2.imread(image_path)
    # cv2.imshow("aligned", align_img(image))
    # cv2.waitKey(0)
    prepared_image = prepare_img(image)
    question_contours = generate_bubble_contours(prepared_image)
    # objects = identify_object_contours(question_contours)

    image_with_contours = image.copy()

    image_with_bubble = image.copy()

    cv2.drawContours(image_with_contours, question_contours, -1, (255, 255, 0), 2)

    filled_in = identify_bubbled(image, question_contours)

    cv2.drawContours(image_with_bubble, filled_in, -1, (0, 255, 0), 2)
    cv2.imshow("Bubbled Image", cv2.resize(image_with_bubble, (1080, 900)))

    cv2.imshow("Prepared Image", cv2.resize(prepared_image, (800, 800)))
    cv2.imshow("Contoured Image", cv2.resize(image_with_contours, (900, 900)))
    # cv2.imshow("Objects Image", cv2.resize(image_with_objects_identified, (900, 900)))

    cv2.imshow("Contoured Image", image_with_contours)
    cv2.waitKey(0)
