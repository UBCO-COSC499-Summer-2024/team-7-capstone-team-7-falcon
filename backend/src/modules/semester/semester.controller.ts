import {
  Body,
  Controller,
  Get,
  HttpStatus,
  Param,
  ParseIntPipe,
  Patch,
  Post,
  Res,
  UseGuards,
  ValidationPipe,
} from '@nestjs/common';
import { AuthGuard } from '../../guards/auth.guard';
import { SystemRoleGuard } from '../../guards/system-role.guard';
import { SemesterDto } from './dto/semester.dto';
import { Roles } from '../../decorators/roles.decorator';
import { UserRoleEnum } from '../../enums/user.enum';
import { SemesterService } from './semester.service';
import { Response } from 'express';
import {
  SemesterCreationException,
  SemesterNotFoundException,
} from '../../common/errors';
import { ERROR_MESSAGES } from '../../common';

@Controller('semester')
export class SemesterController {
  /**
   * Constructor of SemesterController
   * @param semesterService {SemesterService} - The semester service
   */
  constructor(private readonly semesterService: SemesterService) {}

  /**
   * Get all semesters
   * @param res {Response} - The response object
   * @returns {Promise<Response>} - The response object with semesters or error message
   */
  @UseGuards(AuthGuard)
  @Get('/limited/all')
  async getAllLimited(@Res() res: Response): Promise<Response> {
    const semesters = await this.semesterService.getAllSemestersLimited();

    if (!semesters) {
      return res.status(HttpStatus.NOT_FOUND).send({
        message: ERROR_MESSAGES.semesterController.semestersNotFound,
      });
    }

    return res.status(HttpStatus.OK).send(semesters);
  }

  /**
   * Get all semesters
   * @param res {Response} - The response object
   * @returns {Promise<Response>} - The response object with semesters or error message
   */
  @UseGuards(AuthGuard, SystemRoleGuard)
  @Roles(UserRoleEnum.ADMIN)
  @Get('/all')
  async getAll(@Res() res: Response): Promise<Response> {
    try {
      const semesters = await this.semesterService.getAllSemesters();

      if (semesters.length === 0) {
        return res.status(HttpStatus.NO_CONTENT).send();
      }

      return res.status(HttpStatus.OK).send(semesters);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Create a new semester
   * @param res {Response} - The response object
   * @param body {SemesterCreateDto} - The semester details
   * @returns {Promise<Response>} - The response object
   */
  @UseGuards(AuthGuard, SystemRoleGuard)
  @Roles(UserRoleEnum.ADMIN)
  @Post('/create')
  async post(
    @Res() res: Response,
    @Body(new ValidationPipe()) body: SemesterDto,
  ): Promise<Response> {
    try {
      await this.semesterService.createSemester(body);
      return res.status(HttpStatus.CREATED).send({
        message: 'ok',
      });
    } catch (e) {
      if (e instanceof SemesterCreationException) {
        return res.status(HttpStatus.BAD_REQUEST).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Update semester details by id
   * @param res {Response} - The response object
   * @param sid {number} - The semester id {sid
   * @param body {SemesterDto} - The semester details {body
   * @returns {Promise<Response>} - The response object
   */
  @UseGuards(AuthGuard, SystemRoleGuard)
  @Roles(UserRoleEnum.ADMIN)
  @Patch('/:sid')
  async update(
    @Res() res: Response,
    @Param('sid', ParseIntPipe) sid: number,
    @Body(new ValidationPipe()) body: SemesterDto,
  ): Promise<Response> {
    try {
      await this.semesterService.updateSemester(sid, body);
      return res.status(HttpStatus.NO_CONTENT).send();
    } catch (e) {
      if (e instanceof SemesterNotFoundException) {
        return res.status(HttpStatus.NOT_FOUND).send({
          message: e.message,
        });
      } else if (e instanceof SemesterCreationException) {
        return res.status(HttpStatus.BAD_REQUEST).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Get semester by id
   * @param res {Response} - The response object {res
   * @param sid {number} - The semester id
   * @returns {Promise<Response>} - The response object
   */
  @UseGuards(AuthGuard, SystemRoleGuard)
  @Roles(UserRoleEnum.ADMIN)
  @Get('/:sid')
  async getSemesterById(
    @Res() res: Response,
    @Param('sid', ParseIntPipe) sid: number,
  ): Promise<Response> {
    try {
      const semester = await this.semesterService.getSemesterById(sid);
      return res.status(HttpStatus.OK).send(semester);
    } catch (e) {
      if (e instanceof SemesterNotFoundException) {
        return res.status(HttpStatus.NOT_FOUND).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }
}
