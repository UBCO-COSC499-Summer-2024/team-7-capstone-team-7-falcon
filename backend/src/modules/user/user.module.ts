import { Module, forwardRef } from '@nestjs/common';
import { UserController } from './user.controller';
import { UserService } from './user.service';
import { TokenModule } from '../token/token.module';

@Module({
  controllers: [UserController],
  providers: [UserService],
  exports: [UserService],
  imports: [forwardRef(() => TokenModule)],
})
export class UserModule {}
