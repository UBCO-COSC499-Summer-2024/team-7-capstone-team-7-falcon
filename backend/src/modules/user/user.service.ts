import { Injectable } from '@nestjs/common';
import { UserModel } from './entities/user.entity';
import { OAuthGoogleUserPayload } from '../../common/interfaces';
import {
  EmployeeIdAlreadyExistsException,
  StudentIdAlreadyExistsException,
  UserAlreadyExistsException,
  UserNotFoundException,
} from '../../common/errors';
import { AuthTypeEnum } from '../../enums/user.enum';
import { EmployeeUserModel } from './entities/employee-user.entity';
import { StudentUserModel } from './entities/student-user.entity';
import { UserEditDto } from './dto/user-edit.dto';
import { CourseUserModel } from '../course/entities/course-user.entity';

@Injectable()
export class UserService {
  /**
   * Search for courses based on user id
   * @param user_id {number} - User id
   * @returns {Promise<CourseUserModel[]>} - CourseUserModel[] promise
   */
  public async findUserCoursesById(
    user_id: number,
  ): Promise<CourseUserModel[]> {
    const courses: CourseUserModel[] = await CourseUserModel.find({
      where: { user: { id: user_id }, course: { is_archived: false } },
      relations: ['user', 'course'],
    });

    if (!courses || courses.length === 0) {
      return null;
    }

    return courses;
  }

  /**
   * Returns a user by id
   * @param id {number} - User id
   * @returns {Promise<UserModel>} - User object
   */
  public async getUserById(id: number): Promise<UserModel> {
    const user: UserModel = await UserModel.findOne({
      where: { id },
      relations: ['employee_user', 'student_user'],
    });

    if (!user) {
      return null;
    }

    // Remove password from user object as it is sensitive information
    delete user.password;

    return user;
  }

  /**
   * Returns a user by email
   * @param email {string} - User email
   * @returns {Promise<UserModel>} - User object
   */
  public async findOrCreateUser(
    userPayload: OAuthGoogleUserPayload,
    method: string,
  ): Promise<UserModel> {
    if (method === AuthTypeEnum.GOOGLE_OAUTH) {
      let user: UserModel = await UserModel.findOne({
        where: { email: userPayload.email },
      });

      if (!user) {
        const currentTime = parseInt(new Date().getTime().toString());

        user = new UserModel();
        user.email = userPayload.email;
        user.first_name = userPayload.given_name;
        user.last_name = userPayload.family_name;
        user.avatar_url = userPayload.picture;
        user.auth_type = AuthTypeEnum.GOOGLE_OAUTH;
        user.created_at = currentTime;
        user.updated_at = currentTime;

        await user.save();
      }

      if (user && user.auth_type !== AuthTypeEnum.GOOGLE_OAUTH) {
        throw new UserAlreadyExistsException();
      }

      return user;
    }

    throw new Error('Invalid auth method');
  }

  /**
   * Edit user details
   * @param uid {number} - User id
   * @param userEditBody {UserEditDto} - User edit body
   * @returns {Promise<UserModel>} - User object
   */
  public async editUser(
    uid: number,
    userEditBody: UserEditDto,
  ): Promise<UserModel> {
    const user: UserModel = await UserModel.findOne({
      where: { id: uid },
      relations: ['employee_user', 'student_user'],
    });

    if (!user) {
      throw new UserNotFoundException();
    }

    if (userEditBody.first_name) {
      user.first_name = userEditBody.first_name;
    }

    if (userEditBody.last_name) {
      user.last_name = userEditBody.last_name;
    }

    // Create or update employee number
    if (userEditBody.employee_id) {
      const employeeUser: EmployeeUserModel = await this.findEmployeeNumber(
        userEditBody.employee_id,
      );

      if (employeeUser && employeeUser.user.id !== uid) {
        throw new EmployeeIdAlreadyExistsException();
      }

      let employeeUserRecord: EmployeeUserModel;

      if (!user.employee_user) {
        employeeUserRecord = await EmployeeUserModel.create({
          employee_id: userEditBody.employee_id,
          user,
        }).save();
        user.employee_user = employeeUserRecord;
      } else {
        user.employee_user.employee_id = userEditBody.employee_id;
        await user.employee_user.save();
      }
    }

    // Create or update student number
    if (userEditBody.student_id) {
      const studentUser: StudentUserModel = await this.findStudentNumber(
        userEditBody.student_id,
      );

      if (studentUser && studentUser.user.id !== uid) {
        throw new StudentIdAlreadyExistsException();
      }

      let studentUserRecord: StudentUserModel;

      if (!user.student_user) {
        studentUserRecord = await StudentUserModel.create({
          student_id: userEditBody.student_id,
          user,
        }).save();
        user.student_user = studentUserRecord;
        user.student_user.save();
      } else {
        user.student_user.student_id = userEditBody.student_id;
        user.student_user.save();
      }
    }
    await user.save();

    return user;
  }

  /**
   * Checks if an employee number exists in the database
   * @param employeeId {number} - Employee number
   * @param userId {number} - User id
   * @returns {Promise<EmployeeUserModel>} - Employee user object
   */
  private async findEmployeeNumber(
    employeeId: number,
  ): Promise<EmployeeUserModel> {
    return await EmployeeUserModel.findOne({
      where: { employee_id: employeeId },
      relations: ['user'],
    });
  }

  /**
   * Checks if a student number exists in the database
   * @param studentId {number} - Student number
   * @returns {Promise<StudentUserModel>} - Student user object
   */
  private async findStudentNumber(
    studentId: number,
  ): Promise<StudentUserModel> {
    return await StudentUserModel.findOne({
      where: { student_id: studentId },
      relations: ['user'],
    });
  }
}
