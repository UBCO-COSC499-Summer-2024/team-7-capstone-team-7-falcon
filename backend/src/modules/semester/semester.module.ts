import { Module } from '@nestjs/common';
import { SemesterController } from './semester.controller';
import { SemesterService } from './semester.service';
import { UserService } from '../user/user.service';
import { TokenService } from '../token/token.service';

@Module({
  controllers: [SemesterController],
  providers: [SemesterService, UserService, TokenService],
  exports: [SemesterService],
})
export class SemesterModule {}
