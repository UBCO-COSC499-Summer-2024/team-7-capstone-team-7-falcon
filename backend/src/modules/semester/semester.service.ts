import { Injectable } from '@nestjs/common';
import { SemesterDto } from './dto/semester.dto';
import {
  SemesterCreationException,
  SemesterNotFoundException,
} from '../../common/errors';
import { ERROR_MESSAGES } from '../../common';
import { SemesterModel } from './entities/semester.entity';
import { MoreThan } from 'typeorm';
import { CourseModel } from '../course/entities/course.entity';

@Injectable()
export class SemesterService {
  private readonly TWO_DAYS = 2 * 24 * 60 * 60 * 1000;
  private readonly THREE_MONTHS = 3 * 30 * 24 * 60 * 60 * 1000;

  /**
   * Get all semesters
   * @returns {Promise<SemesterModel[]>} - The promise object
   */
  public async getAllSemestersLimited(): Promise<SemesterModel[]> {
    const dateThreeMonthsAgo: number =
      parseInt(new Date().getTime().toString()) - this.THREE_MONTHS;

    const result = await SemesterModel.find({
      where: { starts_at: MoreThan(dateThreeMonthsAgo) },
      select: ['id', 'name'],
    });

    if (!result || result.length === 0) {
      return null;
    }

    return result;
  }

  /**
   * Get all semesters and their course count
   * @returns {Promise<SemesterModel[]>} - The promise object
   */
  public async getAllSemesters(): Promise<SemesterModel[]> {
    const semesters = await SemesterModel.createQueryBuilder('semester')
      .loadRelationCountAndMap('semester.course_count', 'semester.courses')
      .getMany();

    return semesters;
  }

  /**
   * Create a new semester
   * @param semesterDetails {SemesterDto} - The semester details
   * @returns {Promise<void>} - The promise object
   */
  public async createSemester(semesterDetails: SemesterDto): Promise<void> {
    if (semesterDetails.starts_at >= semesterDetails.ends_at) {
      throw new SemesterCreationException(
        ERROR_MESSAGES.semesterController.semesterStartDateMustBeBeforeEndDate,
      );
    }

    const currentDate: number = parseInt(new Date().getTime().toString());

    if (semesterDetails.starts_at <= currentDate - this.TWO_DAYS) {
      throw new SemesterCreationException(
        ERROR_MESSAGES.semesterController.semesterStartDateMustBeTwoDaysAhead,
      );
    }

    await SemesterModel.create({
      ...semesterDetails,
      created_at: currentDate,
      updated_at: currentDate,
    }).save();

    return;
  }

  /**
   * Update a semester
   * @param sid {number} - The semester id
   * @param semesterDetails {SemesterDto} - The semester details
   * @returns {Promise<void>} - The promise object
   */
  public async updateSemester(
    sid: number,
    semesterDetails: SemesterDto,
  ): Promise<void> {
    const semester = await SemesterModel.findOne({ where: { id: sid } });

    if (!semester) {
      throw new SemesterNotFoundException();
    }

    if (semesterDetails.starts_at >= semesterDetails.ends_at) {
      throw new SemesterCreationException(
        ERROR_MESSAGES.semesterController.semesterStartDateMustBeBeforeEndDate,
      );
    }

    await SemesterModel.update(sid, {
      ...semesterDetails,
      updated_at: parseInt(new Date().getTime().toString()),
    });

    return;
  }

  /**
   * Get a semester by id
   * @param sid {number} - The semester id
   * @returns {Promise<SemesterModel>} - The promise object
   */
  public async getSemesterById(sid: number): Promise<SemesterModel> {
    const semester = await SemesterModel.findOne({ where: { id: sid } });

    if (!semester) {
      throw new SemesterNotFoundException();
    }

    return semester;
  }

  /**
   * Delete a semester
   * @param sid {number} - The semester id
   * @returns {Promise<void>} - The promise object
   */
  public async deleteSemester(sid: number): Promise<void> {
    const semester = await this.getSemesterById(sid);

    await CourseModel.update({ semester: semester }, { semester: null });

    await semester.remove();
  }
}
