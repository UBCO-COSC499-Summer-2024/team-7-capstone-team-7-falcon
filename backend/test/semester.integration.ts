import { UserModel } from '../src/modules/user/entities/user.entity';
import { SemesterModel } from '../src/modules/semester/entities/semester.entity';
import { SemesterModule } from '../src/modules/semester/semester.module';
import { setUpIntegrationTests, signJwtToken } from './utils/testUtils';
import { UserRoleEnum } from '../src/enums/user.enum';

describe('Semester Integration', () => {
  const supertest = setUpIntegrationTests(SemesterModule);

  beforeEach(async () => {
    await UserModel.delete({});
    await UserModel.query('ALTER SEQUENCE user_model_id_seq RESTART WITH 1');

    await SemesterModel.delete({});
    await SemesterModel.query(
      'ALTER SEQUENCE semester_model_id_seq RESTART WITH 1',
    );
  });

  describe('GET /semester/all', () => {
    it('should return status 401 if no token is provided', async () => {
      await supertest().get('/semester/all').expect(401);
    });

    it('should return 404 if no semesters are found', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await supertest()
        .get('/semester/all')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .expect(404)
        .expect({
          message: 'Semesters not found',
        });
    });

    it('should return 200 if semesters are found', async () => {
      const currentDate = parseInt(new Date().getTime().toString());

      await SemesterModel.create({
        name: 'Test Semester',
        created_at: currentDate,
        updated_at: currentDate,
        starts_at: currentDate,
        ends_at: currentDate,
      }).save();

      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await supertest()
        .get('/semester/all')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .expect(200)
        .expect([
          {
            id: 1,
            name: 'Test Semester',
            starts_at: currentDate.toString(),
            ends_at: currentDate.toString(),
            created_at: currentDate.toString(),
            updated_at: currentDate.toString(),
          },
        ]);
    });
  });

  describe('POST /semester/create', () => {
    it('should return status 401 if no token is provided', async () => {
      await supertest().post('/semester/create').expect(401);
    });

    it('should return 401 if the user is not an admin', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await supertest()
        .post('/semester/create')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .expect(401);
    });

    it('should return 400 if one of the required fields is missing', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        role: UserRoleEnum.ADMIN,
      }).save();

      await supertest()
        .post('/semester/create')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .send({
          name: 'Semester 1',
          starts_at: parseInt(new Date().getTime().toString()),
        })
        .expect(400)
        .expect({
          message: [
            'End date is required',
            'ends_at must be a number conforming to the specified constraints',
          ],
          error: 'Bad Request',
          statusCode: 400,
        });
    });

    it('should return 400 if the start date is after the end date', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        role: UserRoleEnum.ADMIN,
      }).save();

      await supertest()
        .post('/semester/create')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .send({
          name: 'Semester 1',
          starts_at: parseInt(new Date().getTime().toString()),
          ends_at: parseInt(new Date().getTime().toString()) - 1000,
        })
        .expect(400)
        .expect({
          message: 'Start date must be before end date',
        });
    });

    it('should return 201 if the semester is created successfully', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        role: UserRoleEnum.ADMIN,
      }).save();

      await supertest()
        .post('/semester/create')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .send({
          name: 'Semester 1',
          starts_at: parseInt(new Date().getTime().toString()),
          ends_at: parseInt(new Date().getTime().toString()) + 1000,
        })
        .expect(201)
        .expect({
          message: 'ok',
        });
    });
  });
});
