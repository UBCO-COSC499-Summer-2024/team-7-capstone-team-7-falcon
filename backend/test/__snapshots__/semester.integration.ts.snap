// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Semester Integration GET /semester/:sid should return 200 if the semester is found 1`] = `
{
  "created_at": "1000000",
  "ends_at": "1000000",
  "id": 1,
  "name": "Test Semester",
  "starts_at": "1000000",
  "updated_at": "1000000",
}
`;

exports[`Semester Integration GET /semester/all should return 200 if semesters are found 1`] = `
[
  {
    "course_count": 10,
    "created_at": "1000000000",
    "ends_at": "1000000000",
    "id": 1,
    "name": "Test Semester",
    "starts_at": "1000000000",
    "updated_at": "1000000000",
  },
]
`;
